File: 2024/day02/Main.scala
import scala.io.Source
import scala.math.{signum, abs}

object Main extends App {
  val numbers = inputNumbers()
  val safeLines = numbers.count(safeLine)
  val safeLinesTolerant = numbers.count(safeLineTolerant)
  // numbers.filter(safeLine).map(println)
  // println(safeLine(List(17, 18, 11, 8, 3)))
  println(s"part(a) | safeLines: ${safeLines}")
  println(s"part(b) | safeLinesTolerant: ${safeLinesTolerant}")

  def inputNumbers(): List[List[Int]] = {
    val fileName = "input.txt"
    Source.fromFile(fileName).getLines()
          .map(_.split("\\s+").map(_.toInt).toList).toList
  }

  def safeLineTolerant(line: List[Int]): Boolean = {
    val linesWithoutOneLevel = line.zipWithIndex.map {
      case (_, idx) => line.patch(idx, Nil, 1)
    }
    linesWithoutOneLevel.exists(safeLine)
  }

  def safeLine(line: List[Int]): Boolean = {
    val checksumTarget = line.size - 1
    val checksum = line.zip(line.tail).map {
      (a, b) => if (abs(a - b) <= 3) signum(a - b) else 0
    }.sum()
    abs(checksum) == checksumTarget
  }
}

File: 2024/day02/main.py
def signum(x):
    # holy christ: why python don't have this no stdlib?!
    return (x > 0) - (x < 0)


def safe_line(line: list[int]) -> bool:
    checksum_target = len(line) - 1
    checksum = [
        signum(a - b) if abs(a - b) <= 3 else 0
        for a, b in zip(line, line[1:])
    ]
    return abs(sum(checksum)) == checksum_target


def main():
    lines = [list(map(int, line.split()))
             for line in open("input.txt").readlines()]
    print("part(a) | safe_lines: ", len(list(filter(safe_line, lines))))


if __name__ == "__main__":
    main()
